pkg/kubelet/server/stats/volume_stat_calculator.go

```diff
@@ -142,7 +143,10 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 		if err != nil {
 			// Expected for Volumes that don't support Metrics
 			if !volume.IsNotSupported(err) {
-				klog.V(4).InfoS("Failed to calculate volume metrics", "pod", klog.KObj(s.pod), "podUID", s.pod.UID, "volumeName", name, "err", err)
+				if oldErr, found := errors[name]; !found || oldErr.Error() != err.Error() {
+					klog.InfoS("Failed to calculate volume metrics", "pod", klog.KObj(s.pod), "podUID", s.pod.UID, "volumeName", name, "err", err)
+					errors[name] = err
```

Comment:
I don't expect duplicate volume names so I think this won't reduce the amount of logs.

@msau42, in the issue, you said "We may need to think about how to not log every time to avoid log spam", do you mean we should consolidate this error for multiple volumes with the same error message?

Reply:
@xing-yang 

Please refer to the [following code] (https://github.com/draveness/kubernetes/blob/98d3738f6d3e140d193bb3fc5384d93b92fbd562/pkg/kubelet/server/stats/fs_resource_analyzer.go#L73C1-L95C2) where this `calcAndStoreStats func()` get used. It is used in `fsResourceAnalyzer` and it call `StartOnce() func` below. So basically `calcAndStoreStats func()` will be called again and again until that pod been deleted, which means if there is an error during volume calculation, that error will be logged multiple times with the same content. I think that's the real place we want to avoid, same content multiple logs.

func (s *volumeStatCalculator) StartOnce() *volumeStatCalculator {
        s.startO.Do(func() {
                go wait.JitterUntil(func() {
                        s.calcAndStoreStats()
                }, s.jitterPeriod, 1.0, true, s.stopChannel)
        })
        return s
}

The current implementation store `errors := map[string]error` and only log the error at the first time and store it in the map. Next time when we do the calculation again, since it was stored in the map, it will not be logged again.

Reply:
I see.  That makes sense.

Reply:
What is `name`? Is it unique across pods?